src/python/turicreate/toolkits/_image_feature_extractor.py

```diff
@@ -82,83 +82,108 @@ def __init__(self, ptModel):
+        state = {}
+        state['num_started'] = 0       # Images read from the SFrame
+        state['num_processed'] = 0     # Images processed by TensorFlow
+        state['total'] = len(dataset)  # Images in the dataset
 
-        # Setup the MXNet model
-        model = MXFeatureExtractor._get_mx_module(self.ptModel.mxmodel,
-                self.data_layer, self.feature_layer, self.context, self.image_shape, batch_size)
+        # We should be using SArrayBuilder, but it doesn't accept ndarray yet.
+        #out = _tc.SArrayBuilder(dtype = array.array)
+        state['out'] = _tc.SArray(dtype=array.array)
 
-        out = _tc.SArrayBuilder(dtype = array.array)
-        progress = { 'num_processed' : 0, 'total' : len(dataset) }
         if verbose:
             print("Performing feature extraction on resized images...")
 
-        # Encapsulates the work done by the MXNet model for a single batch
+        # Provide an iterator-like interface around the SFrame.
+        def has_next_batch():
+            return state['num_started'] < state['total']
+
+        # Yield a numpy array representing one batch of images.
+        def next_batch():
+            # Compute the range of the SFrame to yield.
+            start_index = state['num_started']
+            end_index = min(start_index + batch_size, state['total'])
+            state['num_started'] = end_index
+
+            # Allocate a numpy array with the desired shape.
+            # TODO: Recycle allocations?
+            num_images = end_index - start_index
+            shape = (num_images,) + self.ptModel.input_image_shape
+            batch = _np.zeros(shape, dtype=_np.float32)
+
+            # Resize and load the images.
+            _tc.extensions.sframe_load_to_numpy(image_sf, batch.ctypes.data,
+                                                batch.strides, batch.shape,
+                                                start_index, end_index)
+
+            # TODO: Converge to NCHW everywhere.
+            batch = batch.transpose(0, 2, 3, 1)  # NCHW -> NHWC
+
+            if self.ptModel.input_is_BGR:
+                batch = batch[:,:,:,::-1]  # RGB -> BGR
+
+            return batch
+
+        # Encapsulates the work done by TensorFlow for a single batch.
         def handle_request(batch):
-            model.forward(batch)
-            mx_out = [array.array('d',m) for m in model.get_outputs()[0].asnumpy()]
-            if batch.pad != 0:
-                # If batch size is not evenly divisible by the length, it will loop back around.
-                # We don't want that.
-                mx_out = mx_out[:-batch.pad]
-            return mx_out
-
-        # Copies the output from MXNet into the SArrayBuilder and emits progress
-        def consume_response(mx_out):
-            out.append_multiple(mx_out)
-
-            progress['num_processed'] += len(mx_out)
+            y = self.model.predict(batch)
+            tf_out = [i[0][0] for i in y]
```

Comment:
Does `y.flatten()` do the same thing? That likely a lot faster than a Python loop.

Reply:
`y` comes out with e.g. shape `(64, 1, 1, 2048)`. Calling `y.flatten()` directly yields a `(131072,)` array. I think this will eventually become a list anyway, by the time we pass to the SArray initializer, so there's not too much harm in doing it now.

I did at least replace this with `[i.flatten() for i in y]`

Reply:
ok